File: cpp/leetcode/407.cpp
class Solution {
public:
    int trapRainWater(vector<vector<int>>& heightMap) {
        if(heightMap.empty()) return 0;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> q;
        int m = heightMap.size(), n = heightMap[0].size();
        vector<vector<int>> visited(m, vector<int>(n, 0));
        int ans = 0, Max = INT_MIN;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(!(i == 0 || i == m-1 || j == 0 || j == n-1)) continue;
                q.push(make_pair(heightMap[i][j], i*n+j));
                visited[i][j] = 1;
            }
        }
        vector<vector<int>> dir{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        while(!q.empty()) {
            auto val = q.top(); q.pop();
            int height = val.first, x = val.second/n, y = val.second%n;
            Max = max(Max, height);
            for(auto d: dir) {
                int x2 = x + d[0], y2 = y + d[1];
                if(x2>=m || x2<0 || y2<0 || y2>=n || visited[x2][y2]) continue;
                visited[x2][y2] = 1;
                if(heightMap[x2][y2] < Max) ans += Max - heightMap[x2][y2];
                q.push(make_pair(heightMap[x2][y2], x2*n+y2));
            }
        }
        return ans;
    }
};

File: python/leetcode/407.py
class Solution(object):
    def trapRainWater(self, heightMap):
        m = len(heightMap)
        if m == 0:
            return 0
        n = len(heightMap[0])
        if n == 0:
            return 0
        visited = [[False for i in range(n)] for j in range(m)]
        from Queue import PriorityQueue
        q = PriorityQueue()
        for i in range(m):
            visited[i][0] = True
            q.put([heightMap[i][0],i,0])
            visited[i][n-1] = True
            q.put([heightMap[i][n-1],i,n-1])
        for j in range(1, n-1):
            visited[0][j] = True
            q.put([heightMap[0][j],0,j])
            visited[m-1][j] = True
            q.put([heightMap[m-1][j],m-1,j])
        S = 0
        while not q.empty():
            cell = q.get()
            for (i, j) in [(1,0), (-1,0), (0,1), (0,-1)]:
                x = cell[1] + i
                y = cell[2] + j
                if x in range(m) and y in range(n) and not visited[x][y]:
                    S += max(0, cell[0] - heightMap[x][y])   # how much water at the cell
                    q.put([max(heightMap[x][y],cell[0]),x,y])
                    visited[x][y] = True
        return S